### PR TITLE
Merge Warp10-ext-Interpolation into WarpLib

### PR DESCRIPTION
**Pull request by randomboolean**
Merging and adapting interpolation extension

### CODE REVIEW

**Review by pi-r-p — APPROVED** [review]

File: warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_3D.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+      if (3 != values.length) {
+        throw new WarpScriptException(getName() + " expects 3 components but got " + values.length);
+      }
+
+      if (null == values[0] || null == values[1] || null == values[2]) {
+        return new Object[] {tick, GeoTimeSerie.NO_LOCATION, GeoTimeSerie.NO_ELEVATION, null};
+      }
+
+      double x = ((Number) values[0]).doubleValue();
+      double y = ((Number) values[1]).doubleValue();
```
Comment (stggn) [inline warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_3D.java:97]: Missing type check

File: warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_3D.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    if (!(o instanceof List)) {
+      throw new WarpScriptException(getName() + " expects a sorted LIST as 1st argument");
+    }
+    List o1 = (List) o;
+
+    // fill x
+    int d1 = o1.size();
+    xval = new double[d1];
+    for (int i = 0; i < d1; i++) {
+      xval[i] = ((Number) o1.get(i)).doubleValue();
```
Comment (stggn) [inline warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_3D.java:208]: Missing type check

File: warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_3D.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    }
+
+    // fill y
+    int d2 = o2.size();
+    yval = new double[d2];
+    for (int i = 0; i < d2; i++) {
+      yval[i] = ((Number) o2.get(i)).doubleValue();
```
Comment (stggn) [inline warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_3D.java:215]: Missing type check

File: warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_3D.java [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+    }
+
+    // fill z
+    int d3 = o3.size();
+    zval = new double[d3];
+    for (int i = 0; i < d3; i++) {
+      zval[i] = ((Number) o3.get(i)).doubleValue();
```
Comment (stggn) [inline warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_3D.java:222]: Missing type check

File: warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_ND.java [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+    public Object apply(WarpScriptStack stack) throws WarpScriptException {
+      Object o = stack.pop();
+      if (!(o instanceof List)) {
+        throw new WarpScriptException(getName() + " expects a LIST");
+      }
+      List l = (List) o;
+
+      double[] point = new double[l.size()];
+      for (int i = 0; i < l.size(); i++) {
+        point[i] = ((Number) l.get(i)).doubleValue();
```
Comment (stggn) [inline warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_ND.java:97]: Missing type check

File: warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_ND.java [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+    @Override
+    public Object apply(Object[] args) throws WarpScriptException {
+      long tick = (long) args[0];
+      long[] locations = (long[]) args[4];
+      long[] elevations = (long[]) args[5];
+      Object[] values = (Object[]) args[6];
+
+      double[] point = new double[values.length];
+      for (int i = 0; i < values.length; i++) {
+        point[i] = ((Number) values[i]).doubleValue();
```
Comment (stggn) [inline warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_ND.java:115]: Missing type check

File: warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_ND.java [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+      for (Object it: (List) o) {
+        if (!(it instanceof Double)) {
+          throw new WarpScriptException(getName() + " expects values of type DOUBLE.");
+        }
+      }
+      List<Double> yList = (List<Double>) o;
+
+      o = stack.pop();
+      if (!(o instanceof List)) {
+        throw new WarpScriptException(getName() + " expects a List of List as first argument.");
```
Comment (stggn) [inline warp10/src/main/java/io/warp10/script/interpolation/INTERPOLATOR_ND.java:243]: double type should be check here

**Review by stggn — CHANGES_REQUESTED** [review]

**Review by stggn — APPROVED** [review]

**Pull request merged**